index.js

```diff
@@ -450,6 +478,101 @@ methods.concat('all').forEach(function (method) {
   }
 })
 
+/**
+ * Add a route to the map with the given path and methods.
+ * @param {Map} routeMap
+ * @param {string} path
+ * @param {Array} methods
+ * @private
+ */
+function addRouteToMap (routeMap, path, methods, options) {
+  if (routeMap.has(path)) {
+    const existingMethods = routeMap.get(path)
+    for (const method of methods) {
+      if (!existingMethods.methods.includes(method)) {
+        existingMethods.methods.push(method)
+      }
+    }
+  } else {
+    const { keys } = pathRegexp.pathToRegexp(path)
+
+    routeMap.set(
+      path,
+      {
+        methods: [...methods],
+        keys,
+        options: { strict: options.strict, caseSensitive: options.caseSensitive }
+      }
+    )
+  }
+}
+
+/**
+ * Normalize a path by removing trailing slashes.
+ * @param {string} path
+ * @return {string} normalized path
+ * @private
+ */
+function normalizePath (path) {
+  if (typeof path !== 'string') {
+    return path
+  }
+
+  if (path.endsWith('/') && path.length > 1) {
+    return path.slice(0, -1)
+  }
+
+  return path
+}
+
+/**
+ * Collect routes from a router stack recursively.
+ *
+ * @param {Array} stack - The router stack to collect routes from
+ * @param {string} prefix - The path prefix to prepend to routes
+ * @param {Map} routeMap - The map to store collected routes
+ * @private
+ */
+function collectRoutes (stack, prefix, routeMap, options) {
+  for (const layer of stack) {
+    // for routes without a .use
+    if (layer.pathPatterns && layer.route) {
+      const methods = Object.keys(layer.route.methods).map((method) => method.toUpperCase())
+
+      if (Array.isArray(layer.pathPatterns)) {
+        for (const pathPattern of layer.pathPatterns) {
+          const fullPath = prefix === '/' ? pathPattern : normalizePath(prefix) + pathPattern
+          addRouteToMap(routeMap, fullPath, methods, options)
+        }
+      } else {
+        const fullPath = prefix === '/' ? layer.pathPatterns : prefix + layer.pathPatterns
+        addRouteToMap(routeMap, fullPath, methods, options)
+      }
+    }
+
+    // for layers with a .use (mounted routers)
+    if (layer.pathPatterns && layer.handle && layer.handle.stack && !layer.route) {
```

Comment:
For `.use`, could it make sense to allow consumers to iterate recursively themselves? It's something that could be added in a follow up, but the MVP could be simply `{ path, method, router }`. Every field could also be optional, I guess, since `path: undefined` with `.use(fn)`, `method: undefined` when `all` is used, and `router: undefined` when no nested router.